### ISSUE TITLE
Improper binding while executing query in neo4j 2.2.2 - bug

### ISSUE THREAD

**Issue by ernestoe** [neo4j/neo4j#4692]
Refer to this [graphgist](http://gist.neo4j.org/?85a5301ff283fd95e19a) which contains steps for reproducing the bug. 

This is what the data looks visually:

![visual data](http://i57.tinypic.com/2ms4h3c.png)

This is the pattern I'm matching:

![duplicate relationship pattern](http://i60.tinypic.com/258nuvb.png)

The following query doesn't return anything in Neo4j 2.1 (expected result) as you can see in the graphgist, however it returns the following while browsing in version 2.2.2
- [Node[0]{name:"Category"},:META_RELATIONSHIP[1]{name:"GROUPED_BY"},Node[1]{name:"Review"},:META_RELATIONSHIP[3]{name:"GROUPED_BY"},Node[1]{name:"Review"}]
- [Node[0]{name:"Category"},:META_RELATIONSHIP[0]{name:"GROUPED_BY"},Node[2]{name:"Ambiguous"},:META_RELATIONSHIP[7]{name:"GROUPED_BY"},Node[2]{name:"Ambiguous"}]
- [Node[2]{name:"Ambiguous"},:META_RELATIONSHIP[8]{name:"GROUPED_BY"},Node[3]{name:"GlobalTemplateCategory"},:META_RELATIONSHIP[11]{name:"GROUPED_BY"},Node[3]{name:"GlobalTemplateCategory"}]
- [Node[1]{name:"Review"},:META_RELATIONSHIP[4]{name:"GROUPED_BY"},Node[3]{name:"GlobalTemplateCategory"},:META_RELATIONSHIP[13]{name:"GROUPED_BY"},Node[3]{name:"GlobalTemplateCategory"}]
- [Node[3]{name:"GlobalTemplateCategory"},:META_RELATIONSHIP[12]{name:"GROUPED_BY"},Node[4]{name:"MappingReview"},:META_RELATIONSHIP[15]{name:"GROUPED_BY"},Node[4]{name:"MappingReview"}]
- [Node[3]{name:"GlobalTemplateCategory"},:META_RELATIONSHIP[10]{name:"GROUPED_BY"},Node[5]{name:"ClassificationReview"},:META_RELATIONSHIP[16]{name:"GROUPED_BY"},Node[5]{name:"ClassificationReview"}]

Let me know if this is the right format for chasing this "bug".

**Comment by cleishm** [comment]
Hi @ernestoe,

Firstly, thankyou for the awesome bug report. It's really helpful to have so much detail.

I've reproduced the issue you've described. I simplified it slightly to:

```
CREATE (a {name: 'a'})-[:RELATES]->(b {name: 'b'}), (b)-[:RELATES]->(a)
```

Followed by:

```
MATCH p=(n)-->()<--(n)
RETURN head(nodes(p)) AS a, last(nodes(p)) AS b
```

I'll pass it along to our team.

Thanks again.

Cheers,
Chris

**Comment by cleishm** [comment]
PS. You can get the correct behaviour by using the `RULE` planner, as per the following query:

```
PLANNER RULE MATCH p=(n)-->()<--(n) WITH head(nodes(p)) AS a, last(nodes(p)) AS b RETURN a, b, a = b AS shouldBeTrue
```

**Comment by ernestoe** [comment]
Will try that, thanks for the tip. Happy to help to this product that has brought me so much joy. I obtained these results from my own data while doing essays around Data Quality checking in neo4j, as shown [here](http://gist.neo4j.org/?f961b2fad01d3cc65522)

**Comment by lutovich** [comment]
Hello @ernestoe,

Thanks again for the bug report.
It should be fixed in the codebase when https://github.com/neo4j/neo4j/pull/4725 goes in.
I expect this fix to be released with 2.2.3 and next 2.3 milestone.
